src/qibolab/transpilers/general_connectivity.py

```diff
@@ -0,0 +1,411 @@
+            init_samples (int): Number of random qubit mapping samples to be tested
+            (required only for "greedy" initialization).
+        """
+        if init_samples is not None:
+            self.init_method = QubitInitMethod.greedy
+        self._init_samples = init_samples
+
+    def translate_circuit(self, qibo_circuit):
+        """Translate qibo circuit into a list of two qubit gates to be used by the transpiler.
+
+        Args:
+            qibo_circuit (qibo.Circuit): circuit to be transpiled.
+        """
+        translated_circuit = []
+        for index, gate in enumerate(qibo_circuit.queue):
+            if len(gate.qubits) == 2:
+                gate_qubits = list(gate.qubits)
+                gate_qubits.sort()
+                gate_qubits.append(index)
+                translated_circuit.append(gate_qubits)
+            if len(gate.qubits) >= 3:
+                raise_error("ERROR do not use gates acting on more than 2 qubits")
+        self._circuit_repr = translated_circuit
+
+    def reduce(self, graph):
+        """Reduce the circuit, delete a 2-qubit gate if it can be applied on the current configuration.
+
+        Args:
+            graph (networkx.Graph): current hardware qubit mapping.
+
+        Returns:
+            new_circuit (list): reduced circuit.
+        """
+        new_circuit = deepcopy(self._circuit_repr)
+        while new_circuit != [] and (new_circuit[0][0], new_circuit[0][1]) in graph.edges():
+            del new_circuit[0]
+        return new_circuit
+
+    def subgraph_init(self):
+        """Subgraph isomorphism initialization, NP-complete it can take a long time for large circuits."""
+        H = nx.Graph()
+        H.add_nodes_from([i for i in range(self._connectivity.number_of_nodes())])
+        GM = nx.algorithms.isomorphism.GraphMatcher(self._connectivity, H)
+        i = 0
+        H.add_edge(self._circuit_repr[i][0], self._circuit_repr[i][1])
+        while GM.subgraph_is_monomorphic() == True:
+            result = GM
+            i = i + 1
+            H.add_edge(self._circuit_repr[i][0], self._circuit_repr[i][1])
+            GM = nx.algorithms.isomorphism.GraphMatcher(self._connectivity, H)
+            if self._connectivity.number_of_edges() == H.number_of_edges() or i == len(self._circuit_repr) - 1:
+                G = nx.relabel_nodes(self._connectivity, result.mapping)
+                self._graph = G
+                self._mapping = result.mapping
+                return
+        G = nx.relabel_nodes(self._connectivity, result.mapping)
+        self._graph = G
+        self._mapping = result.mapping
+
+    def greedy_init(self):
+        """initialize the circuit with greedy algorithm let a maximum number of 2-qubit
+        gates can be applied without introducing any SWAP gate"""
+        nodes = self._connectivity.number_of_nodes()
+        keys = list(self._connectivity.nodes())
+        values = list(range(nodes))
+        final_mapping = dict(zip(keys, values))
```

Comment:
```suggestion
        final_mapping = {k: i for i, k in enumerate(keys)}
```
Here dictionary comprehension is very useful, because if we had `values`, instead of keys, we could have done `dict(enumerate(values))`, but a "swapping iterator" is a non-trivial object, but it is very simply expressed with tuple unpacking, i.e. `for i, k`, and dictionary comprehension.